pkg/result/processors/formatter.go

```diff
@@ -0,0 +1,143 @@
+package processors
+
+import (
+	"bytes"
+	"fmt"
+	"os"
+	"slices"
+
+	"github.com/golangci/golangci-lint/pkg/config"
+	"github.com/golangci/golangci-lint/pkg/goformatters"
+	"github.com/golangci/golangci-lint/pkg/goformatters/gci"
+	"github.com/golangci/golangci-lint/pkg/goformatters/gofmt"
+	"github.com/golangci/golangci-lint/pkg/goformatters/gofumpt"
+	"github.com/golangci/golangci-lint/pkg/goformatters/goimports"
+	"github.com/golangci/golangci-lint/pkg/lint/linter"
+	"github.com/golangci/golangci-lint/pkg/logutils"
+	"github.com/golangci/golangci-lint/pkg/result"
+)
+
+type Formatter struct {
+	log        logutils.Log
+	cfg        *config.Config
+	formatters []goformatters.Formatter
+}
+
+func NewFormatter(log logutils.Log, cfg *config.Config,
+	enabledLinters map[string]*linter.Config) (*Formatter, error) {
+	p := &Formatter{
+		log: log,
+		cfg: cfg,
+	}
+
+	if _, ok := enabledLinters[gofmt.Name]; ok {
+		p.formatters = append(p.formatters, gofmt.New(cfg.LintersSettings.Gofmt))
+	}
+
+	_, hasGoFumpt := enabledLinters[gofumpt.Name]
+	_, hasGoImports := enabledLinters[goimports.Name]
+
+	switch {
+	// "dynamic" order based on options.
+	case hasGoFumpt && hasGoImports:
+		// if gofumpt has "ModulePath", it will run after goimports.
+		if cfg.LintersSettings.Gofumpt.ModulePath != "" {
+			p.formatters = append(p.formatters,
+				goimports.New(),
+				gofumpt.New(cfg.LintersSettings.Gofumpt, cfg.Run.Go),
+			)
+		} else {
+			// maybe goimports has "LocalPrefixes`, goimports will run after gofumpt.
+			p.formatters = append(p.formatters,
+				gofumpt.New(cfg.LintersSettings.Gofumpt, cfg.Run.Go),
+				goimports.New(),
+			)
+		}
```

Comment:
Just double checking what's suppose to happen here.

If you have three imports:

```go
import (
    "fmt"
    "org/pkg1"
    "github.com/org/pkg2"
)
```

First running `gofumpt` with `ModulePath` resolving to `org`, this would be as follows since it groups modpath with third party.

```go
import (
    "fmt"

    "org/pkg"
    "github.com/org/pkg2"
)
```

And then running `goimports` with `LocalPrefixes` set to `github.com/org/pkg2` it would sort to:

```go
import (
    "fmt"

    "org/pkg"

    "github.com/org/pkg2"
)
```

Would running them in the other order (`goimports` first) result in `goimports` first changing to

```go
import (
    "fmt"
    "org/pkg"

    "github.com/org/pkg2"
)
```

And then `gofumpt` changing to 

```go
import (
    "fmt"

    "org/pkg"
    "github.com/org/pkg2"
)
```

If so, why can't we _always_ run `gofumpt` first? Either it's configured with `ModulePath` and the sorting will happen in two steps, or it's not and the first time it runs it will just leave `org/pkg` with the stdlib imports which is what `goimports` also would do anyway?

Mostly hoping it would be a way to avoid this dynamic way of running them since it feels complex and creates som kind of trial knowledge.

Reply:
> If so, why can't we always run gofumpt first? 

It was to keep the gofumpt style when using `module-path`.
But as the style is not changed by goimport, we can have a fixed order.